### ISSUE TITLE
multi line (long command on narrow terminal window) editing troubles (ITerm2?)

### ISSUE THREAD

**Issue by hvisage** [ohmyzsh/ohmyzsh#6606]
Hi there,

 Using oh-my-zsh in iTerm2, I can't reliably edit long lines on the prompt line, especially in narrow terminals (typically with multiple windows on several servers, but with extra long line same issue).

 With oh-my-zsh, the cursor position is a blind guess after I've up-arrowed to the line line, not to mention the line not properly shown either.

 When I disable oh-my-zsh, those same lines I can reliably see the cursor position on the various lines of the long line.

 How to fix, where to start showing pictures to explain the issues seen?
Anybody else (especially MacOS(X) users) seen these troubles with long, multiline commands, to edit

**Comment by hevisko** [comment]
Okay, this isse seems to be related to an Iterm2 setting. 
After defaults delete com.googlecode.iterm2, the problem disappeared.
Still trying to find the culprit setting though.

**Comment by haf** [comment]
This also seems to happen when typing out a long line; has it go something to do with the theme's escape codes? Seems to happen after #7544 in my case, but I can't see what error that change might introduce.

**Comment by ivanovv** [comment]
Happens to me all the time after certain command length, I have to switch to bash. Really annoying and no clue how to fix. Standard MacOS Terminal app

**Comment by AaronGillBraun** [comment]
Also having this issue with typing out any command longer than the window width. It's funny, because it's only after having switched to a new mac and setting up iTerm and oh-my-zsh am I experiencing this for this first time. I never had a problem before with the same setup on my old laptop...

**Comment by mcornella** [comment]
OK, everybody please provide a screenshot as well as this info:
- zsh version (`echo $ZSH_VERSION`)
- `echo $TERM`
- terminal emulator
- Oh My Zsh theme

Also, try to reproduce it without OMZ, only sourcing the theme: `source $ZSH/theme/$theme.zsh-theme`, being `$theme` your theme used.

**Comment by jonastr** [comment]
I've noticed the same problem. I'm seeing it on two machines, where I supposedly have the same setup (manual replication).

Here's my debug info:

* `$ZSH_VERSION`: 5.3
* `$TERM`: xterm-256color
* terminal emulator: iTerm v3.2.9
* Oh My Zsh theme: `arrow`

I can also reproduce this problem with the default theme (`robbyrussell`). 

I **cannot** reproduce it in the following scenarios:
* Using `agnoster` theme. Didn't try any others, but should be enough to rule out that all themes are affected.
* Sourcing only the affected theme (e.g. `source $ZSH/themes/arrow.zsh-theme`) and leaving OMZ out of the game
* Using Mac's default `Terminal` app with the affected combination (OMZ + theme)

Looking at the above, it looks to me like it's a combination terminal + OMZ + theme (aka prompt?) that makes it break.

**Comment by mcornella** [comment]
Can you say if you have `$COMPLETION_WAITING_DOTS` enabled?

**Comment by jonastr** [comment]
> Can you say if you have `$COMPLETION_WAITING_DOTS` enabled?

I don't think so:

```
[~] echo $COMPLETION_WAITING_DOTS

[~]
```

**Comment by wdpreez1** [comment]
> Can you say if you have `$COMPLETION_WAITING_DOTS` enabled?

Setting COMPLETION_WAITING_DOTS to 'false' fixed it for me. But when I do C-r (bck-i-srch) the history lines are all offset to the right (from the prompt) by six characters.

**Comment by mcornella** [comment]
Can you post your zshrc file?

**Comment by wdpreez1** [comment]
> Can you post your zshrc file?

[Here](https://pastebin.com/ewRMqP7F) we go. I use the same setup on macOS and Ubuntu but the problem only occurs on Ubuntu (18.04 LTS). I use `tmux` must of the time but the problem occurs in a raw terminal too.

**Comment by mcornella** [comment]
Is it possibly the Powerline prompt? Or maybe that you source the agnoster theme as well, there may be a conflict there?

**Comment by wdpreez1** [comment]
> Is it possibly the Powerline prompt? Or maybe that you source the agnoster theme as well, there may be a conflict there?

Powerline or theme can't be excluded, but I use the exact same setup on macOS where it behaves as expected. Without any grounds to substantiate, I am suspecting terminal configuration, which would have different defaults on Ubuntu.

I am not massively motivated (or skillwise equipped) to resolve this as I mainly use macOS, but if I learn something I will post here.

**Comment by wdpreez1** [comment]
> I am not massively motivated (or skillwise equipped) to resolve this as I mainly use macOS, but if I learn something I will post here.

...but I was too curious. Both the theme and Powerline by themselves each respectively causes the problem. Thanks for suggesting. Both put glyphs of sorts in the prompt so when I next feel moved I'll look at that.

**Comment by jolespin** [comment]
> > Can you say if you have `$COMPLETION_WAITING_DOTS` enabled?
> 
> Setting COMPLETION_WAITING_DOTS to 'false' fixed it for me. But when I do C-r (bck-i-srch) the history lines are all offset to the right (from the prompt) by six characters.

Do I just add this into my `~/.bash_profile` ?

**Comment by wdpreez1** [comment]
> Do I just add this into my `~/.bash_profile` ?

Not sure we're on the same page here:  Oh My Zsh is a package for configuring the zsh, so `~/.bash_profile` does not really come into play;  `$COMPLETION_WAITING_DOTS` would be set in `~/.zshrc` or its equivalent if you have things set up differently.

**Comment by marcin-gryszkalis** [comment]
I had same problem on zsh 5.7.1 on KDE/Konsole with tmux (and in tmux only) but it seems that updating to 5.8 fixed problem.

**Comment by negebauer** [comment]
I'm on zsh `zsh 5.8` and experiencing this problem with oh my zsh without any configuration

**Comment by negebauer** [comment]
Changing the locale fixed it for me https://github.com/ohmyzsh/ohmyzsh/issues/7558#issuecomment-460812339

**Comment by mcapinha** [comment]
I'm using a custom theme (copied from someone else who I've long forgotten) that draws a lightning bolt when a git directory is dirty. In my case, that was the cause of the multi-line history editing problem.

![image](https://user-images.githubusercontent.com/15359207/79451947-44631480-7fdf-11ea-9d99-af87e3eba800.png)

Unchecking "Use Unicode version 9+ widths", in iTerm2, fixed it for me.

**Comment by noio** [comment]
I did what @hevisko said and deleted the iTerm settings, that fixed the problem.

Then I started restoring my desired setup, and I started having the problem again when I toggled the `Minimal` theme for iTerm2. After I do this, long lines start overwriting the same line they were started on, and the cursor is all over teh place.

**Comment by sm-paul-schuette** [comment]
> **Edit the third**: I think it was due to:
> 
> > There are probably color codes in the your PROMPT config which aren't surrounded by %{..%}
> 
> From the other issue thread

That fixed it for me also! Thank you for pointing out.

**Comment by ivanovv** [comment]
for me it was the opposite – non color code input surrounded by `%{%}`

This is the original file https://gist.github.com/BaylorRae/1499367

I changed 

```
  colored_time+="%{$minutes_since_last_commit%}m"
```

to 

```
  colored_time+="$minutes_since_last_commit%1{m%}"
```

and my prompt is healthy now.

**Comment by DFineNormal** [comment]
> Changing the locale fixed it for me [#7558 (comment)](https://github.com/ohmyzsh/ohmyzsh/issues/7558#issuecomment-460812339)

Same fix for a PC running build Win 10 build 10.0.18363

Windows Terminal 1.0.1401 to GNU/Linux synology_denverton_1618+

Avit was broken for me, Agnoster was ok

I didn't try any others.

```shell
# my env
echo $ZSH_VERSION
5.7.1
echo $TERM
xterm-256color

# PROMPT BROKEN
locale
# output
LANG=
LC_CTYPE="en_US.UTF-8"
LC_NUMERIC="en_US.UTF-8"
LC_TIME="en_US.UTF-8"
LC_COLLATE="en_US.UTF-8"
LC_MONETARY="en_US.UTF-8"
LC_MESSAGES="en_US.UTF-8"
LC_PAPER="en_US.UTF-8"
LC_NAME="en_US.UTF-8"
LC_ADDRESS="en_US.UTF-8"
LC_TELEPHONE="en_US.UTF-8"
LC_MEASUREMENT="en_US.UTF-8"
LC_IDENTIFICATION="en_US.UTF-8"
LC_ALL=

# PROMPT OK
locale
# output
LANG=en_US.UTF-8
LC_CTYPE="en_US.UTF-8"
LC_NUMERIC="en_US.UTF-8"
LC_TIME="en_US.UTF-8"
LC_COLLATE="en_US.UTF-8"
LC_MONETARY="en_US.UTF-8"
LC_MESSAGES="en_US.UTF-8"
LC_PAPER="en_US.UTF-8"
LC_NAME="en_US.UTF-8"
LC_ADDRESS="en_US.UTF-8"
LC_TELEPHONE="en_US.UTF-8"
LC_MEASUREMENT="en_US.UTF-8"
LC_IDENTIFICATION="en_US.UTF-8"
LC_ALL=en_US.UTF-8
```

**Comment by cardoso-neto** [comment]
I don't even use zsh, but I was messing around with my prompt, so it'd display the branch name of the git repo I'm in.
Everything worked. I love it.
But I wanted to make it purple for reasons.
So I did this:
```bash
PS1='${debian_chroot:+($debian_chroot)}\[\033[01;32m\]\u@\h\[\033[00m\]:\[\033[01;34m\]\w\[\033[00m\] \e[0;35m$(__git_ps1 "(%s)")\e[m\$' 
```
More specifically:
```bash
\e[0;35m$(__git_ps1 "(%s)")\e[m\$
```
Turns out those escape sequences were wreaking havoc on my multiline editing.
After replacing it with just
```bash
$(__git_ps1 "(%s)")\$
```
everything went back to normal.
